File: web-app/backend/app/schemas/portfolio.py
"""Pydantic response models for /portfolio endpoints."""

from datetime import date
from typing import Literal

from pydantic import BaseModel, Field


class PortfolioPeriod(BaseModel):
    """The rolling window the KPIs cover."""

    start_date: date
    end_date: date
    days: int = Field(default=30, description="Length of the rolling window in days.")


class KPITile(BaseModel):
    """One numeric value with a delta vs the prior period."""

    value: float = Field(description="Raw numeric value for the current period.")
    unit: str = Field(description="Unit label, e.g. kWh, EUR, kg CO2e.")
    delta_pct: float | None = Field(
        default=None,
        description="Percent change vs prior period; None if prior period had no data.",
    )
    direction: Literal["up", "down", "neutral"] = Field(
        default="neutral",
        description="Sign of the delta (physical: rose/fell). UI maps it to good/bad per metric.",
    )


class SolarKPIs(BaseModel):
    """On-site solar tiles. Only present when the portfolio has PV (else null).

    For these metrics 'up' is GOOD (more generation / higher renewable share /
    more CO2 avoided), the opposite of energy/cost/CO2.
    """

    generated: KPITile
    renewable_rate: KPITile
    exported: KPITile
    co2_avoided: KPITile


class PortfolioKPIs(BaseModel):
    """Four top-of-page KPI tiles + the period they cover, plus optional solar."""

    period: PortfolioPeriod
    total_energy: KPITile
    avg_eui: KPITile
    total_cost: KPITile
    total_co2: KPITile
    solar: SolarKPIs | None = None


class PortfolioBuildingRow(BaseModel):
    """One row in the buildings table on /portfolio."""

    fabric_building_id: str
    name: str
    city: str
    country: str
    building_type: str
    floor_area_m2: float
    epc_class: str | None

    kwh_30d: float
    cost_30d_eur: float
    co2_30d_kg: float
    eui_kwh_m2_yr: float | None = Field(
        default=None,
        description="Annualized EUI (kWh/m2/yr). None when floor area is zero or missing.",
    )

    open_anomalies: int = Field(description="Unresolved HIGH+CRITICAL anomalies.")
    open_recommendations: int = Field(
        description="Total recommendations on file. V1: no status filter."
    )

    has_pv: bool
    has_battery: bool
    has_iot: bool
    subscription_tier: str = Field(description="Monitor / Insight / Copilot")


class PortfolioBuildingsResponse(BaseModel):
    """Building table payload."""

    buildings: list[PortfolioBuildingRow]

File: web-app/frontend/lib/api/portfolio.ts
/**
 * Server-only fetch helpers for backend /portfolio endpoints.
 * Types must stay in lockstep with backend app/schemas/portfolio.py.
 */

export type Direction = "up" | "down" | "neutral"

export type KPITile = {
  value: number
  unit: string
  delta_pct: number | null
  direction: Direction
}

export type PortfolioPeriod = {
  start_date: string
  end_date: string
  days: number
}

export type SolarKPIs = {
  generated: KPITile
  renewable_rate: KPITile
  exported: KPITile
  co2_avoided: KPITile
}

export type PortfolioKPIs = {
  period: PortfolioPeriod
  total_energy: KPITile
  avg_eui: KPITile
  total_cost: KPITile
  total_co2: KPITile
  solar: SolarKPIs | null
}

export type PortfolioBuildingRow = {
  fabric_building_id: string
  name: string
  city: string
  country: string
  building_type: string
  floor_area_m2: number
  epc_class: string | null
  epc_year?: number | null
  kwh_30d: number
  cost_30d_eur: number
  co2_30d_kg: number
  eui_kwh_m2_yr: number | null
  open_anomalies: number
  open_recommendations: number
  has_pv: boolean
  has_battery: boolean
  has_iot: boolean
  subscription_tier: string
}

export type PortfolioBuildingsResponse = {
  buildings: PortfolioBuildingRow[]
}

type FetchResult<T> = { ok: true; data: T } | { ok: false; error: string }

function getBackendUrl(): string | null {
  return process.env.BACKEND_URL || null
}

function logError(scope: string, status: number, body: string) {
  console.error(`[${scope}] Backend ${status}: ${body.slice(0, 200)}`)
}

async function getJson<T>(
  scope: string,
  path: string,
  accessToken: string
): Promise<FetchResult<T>> {
  const backendUrl = getBackendUrl()
  if (!backendUrl) {
    console.error(`[${scope}] BACKEND_URL not configured`)
    return { ok: false, error: "Server misconfigured" }
  }
  if (!accessToken) {
    return { ok: false, error: "Missing access token" }
  }

  try {
    const res = await fetch(`${backendUrl}${path}`, {
      method: "GET",
      headers: {
        Authorization: `Bearer ${accessToken}`,
        Accept: "application/json",
      },
      cache: "no-store",
    })

    if (!res.ok) {
      const body = await res.text().catch(() => "<no body>")
      if (res.status === 503 && body.includes("fabric_unavailable")) {
        console.warn(`[${scope}] Fabric data temporarily unavailable`)
        return { ok: false, error: "fabric_unavailable" }
      }
      logError(scope, res.status, body)
      return { ok: false, error: `Backend ${res.status}` }
    }

    const data = (await res.json()) as T
    return { ok: true, data }
  } catch (err) {
    console.error(`[${scope}] Network/unknown error:`, err)
    return { ok: false, error: "Network error" }
  }
}

export async function fetchPortfolioKPIs(
  accessToken: string,
  clientOrgId?: string | null
): Promise<FetchResult<PortfolioKPIs>> {
  const q = clientOrgId ? `?client=${encodeURIComponent(clientOrgId)}` : ""
  return getJson<PortfolioKPIs>("fetchPortfolioKPIs", `/portfolio/kpis${q}`, accessToken)
}

export async function fetchPortfolioBuildings(
  accessToken: string,
  clientOrgId?: string | null
): Promise<FetchResult<PortfolioBuildingsResponse>> {
  const q = clientOrgId ? `?client=${encodeURIComponent(clientOrgId)}` : ""
  return getJson<PortfolioBuildingsResponse>(
    "fetchPortfolioBuildings",
    `/portfolio/buildings${q}`,
    accessToken
  )
}
